toqito/matrix_ops/tests/test_tensor_comb.py

```diff
@@ -1,50 +1,68 @@
 
 
 @pytest.mark.parametrize(
-    "states, k, expected",
+    "k, states, mode, density_matrix, expected_keys, expected_error",
     [
-        # Test tensor_comb for sequence length 1
-        ([e_0, e_1], 1, {(0,): np.outer(e_0, e_0), (1,): np.outer(e_1, e_1)}),
-        # Test tensor_comb for sequence length 2
-        (
-            [e_0, e_1],
-            2,
-            {
-                (0, 0): np.outer(np.kron(e_0, e_0), np.kron(e_0, e_0)),
-                (0, 1): np.outer(np.kron(e_0, e_1), np.kron(e_0, e_1)),
-                (1, 0): np.outer(np.kron(e_1, e_0), np.kron(e_1, e_0)),
-                (1, 1): np.outer(np.kron(e_1, e_1), np.kron(e_1, e_1)),
-            },
-        ),
-        # Test tensor_comb for sequence length 3
-        (
-            [e_0, e_1],
-            3,
-            {
-                (0, 0, 0): np.outer(np.kron(np.kron(e_0, e_0), e_0), np.kron(np.kron(e_0, e_0), e_0)),
-                (0, 0, 1): np.outer(np.kron(np.kron(e_0, e_0), e_1), np.kron(np.kron(e_0, e_0), e_1)),
-                # other possible sequences omitted for brevity...
-            },
-        ),
+        (3,
+        # Injective (default).
+         [np.array([1, 0]), np.array([0, 1])],
+         "injective", False,
+         None,
+         ValueError),
+        # Invalid mode
+        (2,
+         [np.array([1, 0]), np.array([0, 1])],
+         "invalid", False,
+         None,
+         ValueError),
+        # Non-injective mode with k = 1
+        (1,
+         [np.array([1, 0]), np.array([0, 1])],
+         "non-injective", False,
+         [(0,), (1,)],
+         None),
+        # Injective mode with k = 1
+        (1,
+         [np.array([1, 0]), np.array([0, 1])],
+         "injective", False,
+         [(0,), (1,)],
+         None),
+        # Non-injective mode with k=2
+        (2,
+         [np.array([1, 0]), np.array([0, 1])],
+         "non-injective", False,
+         [(0, 0), (0, 1), (1, 0), (1, 1)],
+         None),
     ],
 )
-def test_tensor_comb(states, k, expected):
-    """Test tensor_comb for various sequence lengths."""
-    result = tensor_comb(states, k)
-    for key in expected:
-        assert np.allclose(result[key], expected[key])
+def test_tensor_comb(k, states, mode, density_matrix, expected_keys, expected_error):
+    """Test the tensor_comb function."""
+    if expected_error:
+        with pytest.raises(expected_error):
+            tensor_comb(k, states, mode, density_matrix)
+
+    else:
+        result = tensor_comb(k, states, mode, density_matrix)
+        result_keys = list(result.keys())
+        # check expected keys
+        assert sorted(result.keys()) == sorted(
+            expected_keys)   , f"Expected: {expected_keys}, Got: {list(result.keys())}"
 
+        for seq in expected_keys:
+            state_seq = [states[i] for i in seq]
+            tensor_product = np.array(state_seq[0])
+            for state in state_seq[1:]:
+                tensor_product = np.kron(tensor_product, state)
```

Comment:
What was wrong with the previous?

```
    result = tensor_comb(states, k)
    for key in expected:
        assert np.allclose(result[key], expected[key])
```

Doing thins like this

```
        # check expected keys
        assert sorted(result.keys()) == sorted(
            expected_keys)   , f"Expected: {expected_keys}, Got: {list(result.keys())}"
```

don't make sense as the pytest framework already does something like this.